Add Buildkite configs for tracing and runtime builds.

ScottTodd opened the pull request:
I see this "build configurations" pipeline growing to include other combinations of CMake options we want to test. This pipeline should be able to run on presubmit and mainly checks that configurations are valid and build successfully. I don't have the pipeline configured to run any tests or benchmarks. However, we _could_ set up a "minimum size" configuration and upload some metrics from that to our benchmarking infrastructure.

Sample builds:
* https://buildkite.com/iree/scotttodd-test-pipeline/builds/35
* https://buildkite.com/iree/scotttodd-test-pipeline/builds/37

Additional changes:

* Allow building samples without the compiler (lets the runtime build work without disabling samples)
* Add `CMAKE_DL_LIBS` to `iree::base::tracing` (fixes the tracing build on Linux)
* Rework how options are tested in `iree/samples/simple_embedding/CMakeLists.txt` and how `IREE_HOST_BINARY_ROOT` is used in general. Now, regardless of cross compilation status, if `IREE_HOST_BINARY_ROOT` is set then it will be searched for tool executables. Runtime builds are responsible for ensuring that any host builds were completed using a set of target backends compatible with their HAL drivers (for example: a runtime build `-DIREE_HAL_DRIVERS_TO_BUILD=vmvx` attempting to use a host tools build with `-DIREE_TARGET_BACKENDS_TO_BUILD=vulkan-spirv` will likely fail).

ScottTodd commented on CMakeLists.txt:
```diff
@@ -96,10 +96,6 @@ option(IREE_ENABLE_NEW_INTEGRATION_TESTS "Enables new integration tests and disa
 # Derived flags based on primary options
 #-------------------------------------------------------------------------------
 
-if(${IREE_BUILD_SAMPLES})
-  set(IREE_BUILD_COMPILER ON CACHE BOOL "Build the IREE compiler for sample projects." FORCE)
-endif()
```
Hmm... https://source.cloud.google.com/results/invocations/2a89b97a-30df-4923-a8fd-e922e98b0936/targets/iree%2Fgcp_ubuntu%2Fcmake%2Flinux%2Friscv64%2Fpresubmit/log

https://github.com/google/iree/blob/main/iree/samples/simple_embedding/CMakeLists.txt needs more surgery so https://github.com/google/iree/blob/main/build_tools/cmake/build_riscv.sh builds that sample.

ScottTodd commented:
@hcindyl I made some changes to the CMake configuration for `iree/samples/simple_embedding` in here, including removing the `IREE_BUILD_EMBEDDING_SAMPLES` option. LMK if you'd rather I split those commits into their own PR. I think these changes will be a net simplification while still covering the cases we care about.

hcindyl commented:
> @hcindyl I made some changes to the CMake configuration for `iree/samples/simple_embedding` in here, including removing the `IREE_BUILD_EMBEDDING_SAMPLES` option. LMK if you'd rather I split those commits into their own PR. I think these changes will be a net simplification while still covering the cases we care about.

I am not sure it will cover the cases we want. If I read the logic correctly, `iree/sample/simple_embedding/*` targets are built only when `IREE_BUILD_SAMPLES=ON` and `IREE_HOST_BINARY_ROOT` is set, so it will
* Disable build of `iree/samples/simple_embedding` for the host machine (so we can't test x86_64 machines)
* For bare-metal cross compile config, it can't build the all the samples but `iree/samples/simple_embedding` should be built for testing.

ScottTodd commented:
> > @hcindyl I made some changes to the CMake configuration for `iree/samples/simple_embedding` in here, including removing the `IREE_BUILD_EMBEDDING_SAMPLES` option. LMK if you'd rather I split those commits into their own PR. I think these changes will be a net simplification while still covering the cases we care about.
> 
> I am not sure it will cover the cases we want. If I read the logic correctly, `iree/sample/simple_embedding/*` targets are built only when `IREE_BUILD_SAMPLES=ON` and `IREE_HOST_BINARY_ROOT` is set, so it will
> 
> * Disable build of `iree/samples/simple_embedding` for the host machine (so we can't test x86_64 machines)
> * For bare-metal cross compile config, it can't build the all the samples but `iree/samples/simple_embedding` should be built for testing.

The `OR` in `if(${IREE_HAL_DRIVER_DYLIB} AND (${IREE_TARGET_BACKEND_DYLIB-LLVM-AOT} OR DEFINED IREE_HOST_BINARY_ROOT))` should cover this.

Combined host/runtime build: `IREE_HAL_DRIVER_DYLIB` and `IREE_TARGET_BACKEND_DYLIB-LLVM-AOT` are both set -> sample is built using tools from the same build.
Split host/runtime build: Host tools are built with `IREE_TARGET_BACKEND_DYLIB-LLVM-AOT`, runtime is built with `IREE_HAL_DRIVER_DYLIB` and `IREE_HOST_BINARY_ROOT` -> host tools are used to build the sample.

GMNGeoffrey reviewed (approved):
LGTM assuming Cindy is happy with the samples build settings

ScottTodd commented:
> LGTM assuming Cindy is happy with the samples build settings

We're chatting offline a bit too. I just pushed another commit working around https://github.com/google/iree/issues/6247 for now.

hcindyl commented:
> > LGTM assuming Cindy is happy with the samples build settings
> 
> We're chatting offline a bit too. I just pushed another commit working around #6247 for now.

Tested locally and the branch works for bare-metal. Thanks for the change.
